File: adaptive_repair/data/python_programs/rpn_eval.py
def rpn_eval(tokens):
    def op(symbol, a, b):
        return {
            '+': lambda a, b: a + b,
            '-': lambda a, b: a - b,
            '*': lambda a, b: a * b,
            '/': lambda a, b: a / b
        }[symbol](a, b)

    stack = []

    for token in tokens:
        if isinstance(token, float):
            stack.append(token)
        else:
            # The first operand popped (a) is the right-hand side operand.
            # The second operand popped (b) is the left-hand side operand.
            # For non-commutative operations (like subtraction and division),
            # the operation should be b OP a.
            a = stack.pop() # Right operand
            b = stack.pop() # Left operand
            stack.append(
                op(token, b, a) # Corrected order: b then a
            )

    return stack.pop()



"""
Reverse Polish Notation

Four-function calculator with input given in Reverse Polish Notation (RPN).

Input:
    A list of values and operators encoded as floats and strings

Precondition:
    all(
        isinstance(token, float) or token in ('+', '-', '*', '/') for token in tokens
    )

Example:
    >>> rpn_eval([3.0, 5.0, '+', 2.0, '/'])
    4.0
"""

File: adaptive_repair/data/java_programs/RPN_EVAL.java
package java_programs;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.BiFunction;

public class RPN_EVAL {

    private static final Map<String, BiFunction<Double, Double, Double>> operators = new HashMap<>();

    static {
        operators.put("+", (a, b) -> a + b);
        operators.put("-", (a, b) -> a - b);
        operators.put("*", (a, b) -> a * b);
        operators.put("/", (a, b) -> a / b);
    }

    public static double rpn_eval(List<Object> tokens) {
        Deque<Double> stack = new ArrayDeque<>();

        for (Object token : tokens) {
            if (token instanceof Number) {
                stack.push(((Number) token).doubleValue()); // Ensure numbers are treated as doubles
            } else if (token instanceof String) {
                String operator = (String) token;
                if (!operators.containsKey(operator)) {
                    throw new IllegalArgumentException("Unknown operator: " + operator);
                }

                // For non-commutative operations like subtraction and division,
                // the order of operands matters. In RPN, when an operator is encountered,
                // the top two operands are popped. The first popped operand (top of stack)
                // is the right-hand side (RHS), and the second popped operand (below top of stack)
                // is the left-hand side (LHS).
                // The operation should be LHS op RHS.
                
                if (stack.size() < 2) {
                    throw new IllegalArgumentException("Insufficient operands for operator: " + operator);
                }

                // Pop RHS first
                Double rhs = stack.pop(); 
                // Pop LHS second
                Double lhs = stack.pop(); 

                // Apply the operator as 'lhs op rhs'
                Double result = operators.get(operator).apply(lhs, rhs);
                stack.push(result);
            } else {
                throw new IllegalArgumentException("Invalid token type: " + token.getClass().getName());
            }
        }

        if (stack.size() != 1) {
            throw new IllegalArgumentException("Invalid RPN expression: stack size is " + stack.size() + " at the end.");
        }

        return stack.pop();
    }
}